## Задача 1: 
Скачать данные по ссылке https://www.kaggle.com/datasets/esratmaria/house-price-dataset-with-other-information \
Считать данные с помощью pandas\
Вывести на экран первые 5 строк

In [274]:
import pandas as pd

In [275]:
df = pd.read_csv(r'C:\Users\Geralt\tmp\.env\Scripts\kc_house_data.csv', encoding = 'cp1251')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### 1.1.
Создайте новый признак price_per_sq_lot, который будет содержать стоимость за 1 кв. метр общей площади.

In [276]:
df['price_per_sq_lot'] = df['price'] / df['sqft_lot']
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      1180.0              0   
1      7242     2.0           0     0  ...      2170.0            400   
2     10000     1.0           0     0  ...       770.0              0   
3      5000     1.0           0     0  ...      1050.0            910   
4      8080     1.0           0     0  ...      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  price_per_sq_lot  
0        5650         39.274336  
1        7639         74.288870  
2        8062         18.000000  
3        5000        120.800000  
4        7503         63.118812  

[5 rows x 22 columns]

### 1.2 
Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома. Если реновации дома не было, то в новом признаке поставьте 0

In [277]:
df['delta_renovated'] = df['yr_renovated'] - df['yr_built']
df.loc[df['delta_renovated'] <0, 'delta_renovated'] = 0
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_basement  \
0             1180      5650     1.0           0     0  ...              0   
1             2570      7242     2.0           0     0  ...            400   
2              770     10000     1.0           0     0  ...              0   
3             1960      5000     1.0           0     0  ...            910   
4             1680      8080     1.0           0     0  ...              0   
...            ...       ...     ...         ...   ...  ...            ...   
21608         1530      1131     3.0           0     0  ...              0   
21609         2310      5813     2.0           0     0  ...              0   
21610         1020      1350     2.0           0     0  ...              0   
21611         1600      2388     2.0           0     0  ...              0   
21612         1020      1076     2.0           0     0  ...              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0          1955             0    98178  47.5112 -122.257           1340   
1          1951          1991    98125  47.7210 -122.319           1690   
2          1933             0    98028  47.7379 -122.233           2720   
3          1965             0    98136  47.5208 -122.393           1360   
4          1987             0    98074  47.6168 -122.045           1800   
...         ...           ...      ...      ...      ...            ...   
21608      2009             0    98103  47.6993 -122.346           1530   
21609      2014             0    98146  47.5107 -122.362           1830   
21610      2009             0    98144  47.5944 -122.299           1020   
21611      2004             0    98027  47.5345 -122.069           1410   
21612      2008             0    98144  47.5941 -122.299           1020   

       sqft_lot15  price_per_sq_lot  delta_renovated  
0            5650         39.274336                0  
1            7639         74.288870               40  
2            8062         18.000000                0  
3            5000        120.800000                0  
4            7503         63.118812                0  
...           ...               ...              ...  
21608        1509        318.302387                0  
21609        7200         68.811285                0  
21610        2007        297.852593                0  
21611        1287        167.504188                0  
21612        1357        302.044610                0  

[21613 rows x 23 columns]

### 1.3 
Создайте признаки года продажи, месяца продажи

In [278]:
df['deal_year'] = df['date'].apply(lambda x: str(x)[0:4])
df['deal_month'] = df['date'].apply(lambda x: str(x)[4:6])
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  yr_renovated  \
0             1180      5650     1.0           0     0  ...             0   
1             2570      7242     2.0           0     0  ...          1991   
2              770     10000     1.0           0     0  ...             0   
3             1960      5000     1.0           0     0  ...             0   
4             1680      8080     1.0           0     0  ...             0   
...            ...       ...     ...         ...   ...  ...           ...   
21608         1530      1131     3.0           0     0  ...             0   
21609         2310      5813     2.0           0     0  ...             0   
21610         1020      1350     2.0           0     0  ...             0   
21611         1600      2388     2.0           0     0  ...             0   
21612         1020      1076     2.0           0     0  ...             0   

       zipcode      lat     long  sqft_living15  sqft_lot15  price_per_sq_lot  \
0        98178  47.5112 -122.257           1340        5650         39.274336   
1        98125  47.7210 -122.319           1690        7639         74.288870   
2        98028  47.7379 -122.233           2720        8062         18.000000   
3        98136  47.5208 -122.393           1360        5000        120.800000   
4        98074  47.6168 -122.045           1800        7503         63.118812   
...        ...      ...      ...            ...         ...               ...   
21608    98103  47.6993 -122.346           1530        1509        318.302387   
21609    98146  47.5107 -122.362           1830        7200         68.811285   
21610    98144  47.5944 -122.299           1020        2007        297.852593   
21611    98027  47.5345 -122.069           1410        1287        167.504188   
21612    98144  47.5941 -122.299           1020        1357        302.044610   

       delta_renovated  deal_year  deal_month  
0                    0       2014          10  
1                   40       2014          12  
2                    0       2015          02  
3                    0       2014          12  
4                    0       2015          02  
...                ...        ...         ...  
21608                0       2014          05  
21609                0       2015          02  
21610                0       2014          06  
21611                0       2015          01  
21612                0       2014          10  

[21613 rows x 25 columns]

### 1.4 
Удалите признаки date, zipcode, lat, long

In [279]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
0      7129300520  221900.0         3       1.00         1180      5650   
1      6414100192  538000.0         3       2.25         2570      7242   
2      5631500400  180000.0         2       1.00          770     10000   
3      2487200875  604000.0         4       3.00         1960      5000   
4      1954400510  510000.0         3       2.00         1680      8080   
...           ...       ...       ...        ...          ...       ...   
21608   263000018  360000.0         3       2.50         1530      1131   
21609  6600060120  400000.0         4       2.50         2310      5813   
21610  1523300141  402101.0         2       0.75         1020      1350   
21611   291310100  400000.0         3       2.50         1600      2388   
21612  1523300157  325000.0         2       0.75         1020      1076   

       floors  waterfront  view  condition  ...  sqft_above  sqft_basement  \
0         1.0           0     0          3  ...      1180.0              0   
1         2.0           0     0          3  ...      2170.0            400   
2         1.0           0     0          3  ...       770.0              0   
3         1.0           0     0          5  ...      1050.0            910   
4         1.0           0     0          3  ...      1680.0              0   
...       ...         ...   ...        ...  ...         ...            ...   
21608     3.0           0     0          3  ...      1530.0              0   
21609     2.0           0     0          3  ...      2310.0              0   
21610     2.0           0     0          3  ...      1020.0              0   
21611     2.0           0     0          3  ...      1600.0              0   
21612     2.0           0     0          3  ...      1020.0              0   

       yr_built  yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  \
0          1955             0           1340        5650         39.274336   
1          1951          1991           1690        7639         74.288870   
2          1933             0           2720        8062         18.000000   
3          1965             0           1360        5000        120.800000   
4          1987             0           1800        7503         63.118812   
...         ...           ...            ...         ...               ...   
21608      2009             0           1530        1509        318.302387   
21609      2014             0           1830        7200         68.811285   
21610      2009             0           1020        2007        297.852593   
21611      2004             0           1410        1287        167.504188   
21612      2008             0           1020        1357        302.044610   

       delta_renovated  deal_year deal_month  
0                    0       2014         10  
1                   40       2014         12  
2                    0       2015         02  
3                    0       2014         12  
4                    0       2015         02  
...                ...        ...        ...  
21608                0       2014         05  
21609                0       2015         02  
21610                0       2014         06  
21611                0       2015         01  
21612                0       2014         10  

[21613 rows x 21 columns]

## Задача 2:
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})

In [280]:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})
clients.head()

client_id    house_id
0       1459  8965450190
1       4684  6823100225
2       3498  5104540330
3       3942  2131701075
4       4535  1522700060

### 2.1 
Присоедините к таблице clients данные по домам через метод join

In [296]:
df_id = df.set_index('id')
clients_id = clients.set_index('house_id')
joined = df_id.join(clients_id).reset_index()
joined.dropna(subset=['client_id']) 
#выведем только те строки, которые к которым присодинилось не пустое id клиента из 2 таблицы
#это 33 строки

index      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
822     339600110   395000.0         3       2.50         1610      3755   
1669    809001400   925000.0         3       1.00         1630      3200   
1750    825069097   770000.0         3       2.50         2650     40705   
2306   1099600250   260000.0         3       1.75         1710      6400   
2602   1189000207   387000.0         2       2.50         1170      1394   
3623   1522700060   518000.0         4       2.75         2520     14021   
3927   1622049154   289900.0         3       1.75         1899     11325   
4446   1788800910   190000.0         3       1.00         1200     10458   
4980   1951600150   180000.0         3       1.00         1610      8500   
5473   2131701075   420000.0         3       1.75         1720      5000   
7250   2768100205   519000.0         4       2.50         1950      2617   
8208   3123049142   452000.0         3       2.25         2600     14810   
8821   3303980210  1115000.0         4       3.75         4040     14212   
10736  3889100029   810000.0         3       2.50         2670     10481   
10996  3995700435   265000.0         4       3.00         1940      8170   
12489  5083000375   170000.0         3       1.00         1310      9529   
12490  5083000375   235000.0         3       1.00         1310      9529   
12672  5104540330   679000.0         4       2.50         3680      7236   
13229  5428000070   770000.0         5       3.50         4750      8234   
14117  6131600255   202500.0         3       2.00         1540      8316   
14397  6306800010   436472.0         4       2.50         2692      8392   
15187  6821600300   886000.0         3       2.25         2380      6000   
15196  6823100225   700000.0         4       1.75         1870      6000   
15538  7137950720   339100.0         4       2.50         2350     10655   
17114  7731100066   545000.0         3       1.00         1510      5000   
17849  7920100025   450000.0         2       1.00          740      5100   
18473  8100400160   700000.0         3       2.25         2330     11424   
18611  8146100580   765000.0         3       1.00         1220      7585   
18787  8563000110   427000.0         4       1.75         1460      9750   
19723  8861700030   510000.0         3       1.50         2400     10275   
19978  8965450190   295000.0         3       2.50         1500      3060   
21062  9510920050   725000.0         3       2.50         2980     16996   
21145  9524100196   239000.0         2       1.50          680       772   

       floors  waterfront  view  condition  ...  sqft_basement  yr_built  \
822       2.0           0     0          3  ...              0      1987   
1669      1.5           0     0          4  ...              0      1912   
1750      2.0           0     0          3  ...              0      1994   
2306      1.0           0     0          4  ...            470      1976   
2602      2.0           0     0          3  ...              0      2001   
3623      2.0           0     0          3  ...              0      1999   
3927      2.0           0     0          3  ...              0      1943   
4446      1.0           0     0          4  ...              0      1961   
4980      1.5           0     0          3  ...              0      1959   
5473      1.5           0     0          3  ...              0      1932   
7250      1.5           0     0          4  ...            700      1910   
8208      1.0           0     2          4  ...           1110      1956   
8821      2.0           0     0          3  ...              0      2002   
10736     2.0           0     0          3  ...              0      2003   
10996     1.0           0     0          4  ...              0      1948   
12489     1.0           0     0          3  ...              0      1956   
12490     1.0           0     0          3  ...              0      1956   
12672     2.0           0     0          3  ...            

### 2.2 
Присоедините к таблице clients данные по домам через метод merge
Это нужно, чтобы понимать, какие дома покупались клиентами
house_id - это индексы датафрейма с домами

In [282]:
merged = df.merge(clients, left_on='id', right_on='house_id')
merged

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   8100400160   700000.0         3       2.25         2330     11424     2.0   
1   7920100025   450000.0         2       1.00          740      5100     1.0   
2    825069097   770000.0         3       2.50         2650     40705     2.0   
3   5083000375   170000.0         3       1.00         1310      9529     1.0   
4   5083000375   235000.0         3       1.00         1310      9529     1.0   
5    339600110   395000.0         3       2.50         1610      3755     2.0   
6   6821600300   886000.0         3       2.25         2380      6000     2.0   
7   1099600250   260000.0         3       1.75         1710      6400     1.0   
8   3889100029   810000.0         3       2.50         2670     10481     2.0   
9   3123049142   452000.0         3       2.25         2600     14810     1.0   
10  8965450190   295000.0         3       2.50         1500      3060     2.0   
11  7731100066   545000.0         3       1.00         1510      5000     1.5   
12  5104540330   679000.0         4       2.50         3680      7236     2.0   
13  1189000207   387000.0         2       2.50         1170      1394     2.0   
14  1622049154   289900.0         3       1.75         1899     11325     2.0   
15   809001400   925000.0         3       1.00         1630      3200     1.5   
16  8861700030   510000.0         3       1.50         2400     10275     1.0   
17  1522700060   518000.0         4       2.75         2520     14021     2.0   
18  3995700435   265000.0         4       3.00         1940      8170     1.0   
19  1788800910   190000.0         3       1.00         1200     10458     1.0   
20  1951600150   180000.0         3       1.00         1610      8500     1.5   
21  2131701075   420000.0         3       1.75         1720      5000     1.5   
22  8146100580   765000.0         3       1.00         1220      7585     1.0   
23  7137950720   339100.0         4       2.50         2350     10655     2.0   
24  6823100225   700000.0         4       1.75         1870      6000     1.0   
25  6131600255   202500.0         3       2.00         1540      8316     1.0   
26  9510920050   725000.0         3       2.50         2980     16996     2.0   
27  8563000110   427000.0         4       1.75         1460      9750     1.0   
28  3303980210  1115000.0         4       3.75         4040     14212     2.0   
29  2768100205   519000.0         4       2.50         1950      2617     1.5   
30  6306800010   436472.0         4       2.50         2692      8392     2.0   
31  5428000070   770000.0         5       3.50         4750      8234     2.0   
32  9524100196   239000.0         2       1.50          680       772     2.0   

    waterfront  view  condition  ...  yr_built  yr_renovated  sqft_living15  \
0            0     0          4  ...      1984             0           2050   
1            0     0          4  ...      1947             0            920   
2            0     0          3  ...      1994             0           2550   
3            0     0          3  ...      1956             0           1330   
4            0     0          3  ...      1956             0           1330   
5            0     0          3  ...      1987             0           1300   
6            0     0          5  ...      1931             0           2000   
7            0     0          4  ...      1976             0           1600   
8            0     0          3  ...      2003             0           2620   
9            0     2          4  ...      1956             0           2450   
10           0     0          3  ...      1994             0           2700   
11           0     0          3  ...      1909             0           1680   
12           0     0          3  ...      2006             0           3310   
13           0     0          3  ...      2001             0           1250   
14           0     0          3  ...      1943          2005           2

_Видна разница работы методов merge и join_

## Задача 3.
Составьте несколько сводных таблиц

### 3.1 
Найдите среднюю стоимость домов в зависимости от количества спален
Отсортируйте от меньшей стоимости к большей

In [283]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price')

price
bedrooms              
1         3.176429e+05
2         4.013727e+05
0         4.095038e+05
3         4.662321e+05
11        5.200000e+05
4         6.354195e+05
33        6.400000e+05
5         7.865998e+05
10        8.193333e+05
6         8.255206e+05
9         8.939998e+05
7         9.511847e+05
8         1.105077e+06

### 3.2 
Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

In [284]:
# Первый вариент через groupby
condition = df[['price', 'condition']].groupby('condition')
condition.agg({'price': ['min', 'mean', 'max']})

price                          
                min           mean        max
condition                                    
1           78000.0  334431.666667  1500000.0
2           80000.0  327287.145349  2555000.0
3           75000.0  542012.578148  7062500.0
4           89000.0  521200.390033  7700000.0
5          110000.0  612418.089359  3650000.0

In [285]:
# Второй вариант через pivot table
df.pivot_table(index = 'condition', values = 'price', aggfunc = ['min', 'mean', 'max'])

min           mean        max
              price          price      price
condition                                    
1           78000.0  334431.666667  1500000.0
2           80000.0  327287.145349  2555000.0
3           75000.0  542012.578148  7062500.0
4           89000.0  521200.390033  7700000.0
5          110000.0  612418.089359  3650000.0

### 3.3 
Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [286]:
pd.crosstab(df.waterfront, df.view)

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

### 3.4 
Каких домов в зависимости от этажности и количества спален больше?

In [287]:
df_c = df[['id', 'floors', 'bedrooms']].groupby(['floors', 'bedrooms']).count()
df_c[df_c['id'] == df_c['id'].max()]

id
floors bedrooms      
1.0    3         5455

Наибольшее количество домов с 1 одним этажом и тремя спальнями - 5455 домов.

## 3.5 
Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [290]:
pd.crosstab(df.condition, df.grade, values = df.price, aggfunc = 'median')

grade            1         3         4         5         6         7   \
condition                                                               
1          142000.0       NaN  150000.0  190000.0  255000.0  403500.0   
2               NaN  280000.0  145000.0  180000.0  235000.0  305000.0   
3               NaN   75000.0  205000.0  234475.0  265000.0  357500.0   
4               NaN       NaN  238525.0  229975.0  295000.0  390000.0   
5               NaN  262000.0       NaN  227450.0  285475.0  456000.0   

grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0        NaN        NaN        NaN        NaN        NaN  
2          429000.0   715000.0  1752500.0        NaN        NaN        NaN  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0        NaN

_Там, где в ячейку попадает количество домов, равное 0, медиана цены пустая, соответсвенно._